src/mariadb/videx/ha_videx.cc

```diff
@@ -0,0 +1,992 @@
+		HTON_NATIVE_SYS_VERSIONING | HTON_WSREP_REPLICATION |
+			HTON_REQUIRES_CLOSE_AFTER_TRUNCATE | HTON_TRUNCATE_REQUIRES_EXCLUSIVE_USE | HTON_REQUIRES_NOTIFY_TABLEDEF_CHANGED_AFTER_COMMIT;
+
+	videx_hton->update_optimizer_costs = videx_update_optimizer_costs;
+	videx_hton->tablefile_extensions= ha_videx_exts;
+
+	DBUG_RETURN(0);
+}
+
+videx_share *ha_videx::get_share()
+{
+	videx_share *tmp_share;
+
+	DBUG_ENTER("ha_videx::get_share()");
+
+	lock_shared_ha_data();
+	if (!(tmp_share= static_cast<videx_share*>(get_ha_share_ptr())))
+	{
+		tmp_share= new videx_share;
+		if (!tmp_share)
+			goto err;
+
+		set_ha_share_ptr(static_cast<Handler_share*>(tmp_share));
+	}
+err:
+	unlock_shared_ha_data();
+	DBUG_RETURN(tmp_share);
+}
+
+static handler* videx_create_handler(handlerton *hton,
+					TABLE_SHARE *table, 
+					MEM_ROOT *mem_root)
+{
+	return new (mem_root) ha_videx(hton, table);
+}
+
+ha_videx::ha_videx(
+	handlerton*	hton,
+	TABLE_SHARE*	table_arg)
+	:handler(hton, table_arg),
+	m_user_thd(),
+	m_int_table_flags(HA_REC_NOT_IN_SEQ
+			  | HA_NULL_IN_KEY
+			  | HA_CAN_VIRTUAL_COLUMNS
+			  | HA_CAN_INDEX_BLOBS
+			  | HA_CAN_SQL_HANDLER
+			  | HA_REQUIRES_KEY_COLUMNS_FOR_DELETE
+			  | HA_PRIMARY_KEY_REQUIRED_FOR_POSITION
+			  | HA_PRIMARY_KEY_IN_READ_INDEX
+			  | HA_BINLOG_ROW_CAPABLE
+			  | HA_CAN_GEOMETRY
+			  | HA_PARTIAL_COLUMN_READ
+			  | HA_TABLE_SCAN_ON_INDEX
+			  | HA_CAN_EXPORT
+        	  | HA_ONLINE_ANALYZE
+			  | HA_CAN_RTREEKEYS
+			  | HA_CAN_TABLES_WITHOUT_ROLLBACK
+			  | HA_CAN_ONLINE_BACKUPS
+			  | HA_CONCURRENT_OPTIMIZE
+			  | HA_CAN_SKIP_LOCKED
+		  ),
+	m_start_of_scan(),
+    m_mysql_has_locked()
+{}
+
+ha_videx::~ha_videx() = default;
+
+const char* ha_videx::table_type() const
+{
+	return "VIDEX";
+}
+
+handler::Table_flags ha_videx::table_flags() const
+{
+	THD*			thd = ha_thd();
+	handler::Table_flags	flags = m_int_table_flags;
+
+	if (thd_sql_command(thd) == SQLCOM_CREATE_TABLE) {
+		flags|= HA_REQUIRE_PRIMARY_KEY;
+	}
+
+	if (thd_tx_isolation(thd) <= ISO_READ_COMMITTED) {
+		return(flags);
+	}
+
+	return(flags | HA_BINLOG_STMT_CAPABLE);
+}
+
+ulong ha_videx::index_flags(
+	uint	key,
+	uint,
+	bool) const
+{
+	if (table_share->key_info[key].algorithm == HA_KEY_ALG_FULLTEXT) {
+		return(0);
+	}
+
+	/* For spatial index, we don't support descending scan
+	and ICP so far. */
+	if (table_share->key_info[key].algorithm == HA_KEY_ALG_RTREE) {
+		return HA_READ_NEXT | HA_READ_ORDER| HA_READ_RANGE
+			| HA_KEYREAD_ONLY | HA_KEY_SCAN_NOT_ROR;
+	}
+
+	ulong flags= key == table_share->primary_key
+		? HA_CLUSTERED_INDEX : HA_KEYREAD_ONLY | HA_DO_RANGE_FILTER_PUSHDOWN;
+
+	flags |= HA_READ_NEXT | HA_READ_PREV | HA_READ_ORDER
+		| HA_READ_RANGE
+		| HA_DO_INDEX_COND_PUSHDOWN;
+	return(flags);
+}
+
+uint ha_videx::max_supported_keys() const
+{
+	return(MAX_KEY);
+}
+
+uint ha_videx::max_supported_key_length() const
+{
+	return(3500);
+}
+
+uint ha_videx::max_supported_key_part_length() const
```

Comment:
Let's add a reference to MariaDB-InnoDB code: We aim to maintain consistency with InnoDB; see `ha_innobase::max_supported_key_length()`.